File: python/examples/image_edge_detection.py
#!/usr/bin/env python3
"""
Image Edge Detection via Frequency Filtering

Demonstrates:
- High-pass filtering to emphasize edges
- Low-pass filtering to remove noise
- Band-pass filtering for specific frequencies
- Edge detection and sharpening operations
- Comparing different cutoff frequencies
"""

import numpy as np
import spectrograms as sg

def create_test_image(size=256):
    """Create a test image with various features."""
    image = np.zeros((size, size))

    # Large rectangle
    image[50:150, 50:150] = 0.8

    # Small circles (approximate)
    y, x = np.meshgrid(np.arange(size), np.arange(size), indexing='ij')

    circle1 = ((x - 180)**2 + (y - 80)**2) < 20**2
    circle2 = ((x - 180)**2 + (y - 180)**2) < 15**2

    image[circle1] = 1.0
    image[circle2] = 0.6

    # Add some noise
    noise = np.random.randn(size, size) * 0.05
    image = np.clip(image + noise, 0, 1)

    return image


def main():
    print("=== Image Edge Detection via Frequency Filtering ===\n")

    # === 1. Create Test Image ===
    print("1. Creating test image with edges and noise...")
    image = create_test_image(256)

    print(f"   Image size: {image.shape}")
    print(f"   Range: [{image.min():.3f}, {image.max():.3f}]")
    print(f"   Mean: {image.mean():.3f}")
    print(f"   Std dev: {image.std():.3f}\n")

    # === 2. Analyze Original Frequency Content ===
    print("2. Analyzing frequency content...")

    original_power = sg.power_spectrum_2d(image)
    total_power = original_power.sum()

    # Split into frequency bands
    nrows, ncols = original_power.shape
    dc_power = original_power[0, 0]
    low_freq = original_power[:10, :6].sum()  # Low frequencies
    mid_freq = original_power[10:30, 6:16].sum() if nrows > 30 else 0
    high_freq = original_power[30:, 16:].sum() if nrows > 30 else 0

    print(f"   Total power: {total_power:.2e}")
    print(f"   DC component: {dc_power:.2e} ({100*dc_power/total_power:.1f}%)")
    print(f"   Low frequencies: {low_freq:.2e} ({100*low_freq/total_power:.1f}%)")
    print(f"   Mid frequencies: {mid_freq:.2e} ({100*mid_freq/total_power:.1f}%)")
    print(f"   High frequencies: {high_freq:.2e} ({100*high_freq/total_power:.1f}%)\n")

    # === 3. Low-Pass Filter (Smoothing) ===
    print("3. Applying low-pass filters (suppress high frequencies)...")

    cutoffs = [0.3, 0.2, 0.1]
    for cutoff in cutoffs:
        smoothed = sg.lowpass_filter(image, cutoff)

        print(f"   Cutoff {cutoff:.1f}:")
        print(f"     Output range: [{smoothed.min():.3f}, {smoothed.max():.3f}]")
        print(f"     Output std: {smoothed.std():.3f} (lower = smoother)")

    print()

    # === 4. High-Pass Filter (Edge Enhancement) ===
    print("4. Applying high-pass filters (emphasize edges)...")

    for cutoff in cutoffs:
        edges = sg.highpass_filter(image, cutoff)

        # High-pass output can have negative values (edges are +/- around zero)
        edge_strength = np.abs(edges).mean()
        max_edge = np.abs(edges).max()

        print(f"   Cutoff {cutoff:.1f}:")
        print(f"     Output range: [{edges.min():.3f}, {edges.max():.3f}]")
        print(f"     Mean |edge|: {edge_strength:.3f}")
        print(f"     Max |edge|: {max_edge:.3f}")

    print()

    # === 5. Edge Detection ===
    print("5. Detecting edges with detect_edges_fft()...")

    edges = sg.detect_edges_fft(image)

    print(f"   Edge map range: [{edges.min():.3f}, {edges.max():.3f}]")
    print(f"   Mean absolute edge: {np.abs(edges).mean():.3f}")
    print(f"   Max edge magnitude: {np.abs(edges).max():.3f}")

    # Count strong edges (threshold at 0.1)
    strong_edges = np.abs(edges) > 0.1
    edge_pixels = strong_edges.sum()
    print(f"   Strong edge pixels (|edge| > 0.1): {edge_pixels} "
          f"({100*edge_pixels/edges.size:.1f}% of image)\n")

    # === 6. Image Sharpening ===
    print("6. Sharpening image with different amounts...")

    sharpen_amounts = [0.5, 1.0, 2.0]
    for amount in sharpen_amounts:
        sharpened = sg.sharpen_fft(image, amount)

        print(f"   Sharpen amount {amount:.1f}:")
        print(f"     Output range: [{sharpened.min():.3f}, {sharpened.max():.3f}]")
        print(f"     Output std: {sharpened.std():.3f} (higher = more contrast)")

        # Measure edge enhancement
        sharp_edges = sg.highpass_filter(sharpened, 0.2)
        edge_strength = np.abs(sharp_edges).mean()
        print(f"     Edge strength: {edge_strength:.3f}")

    print()

    # === 7. Band-Pass Filter ===
    print("7. Applying band-pass filters (isolate frequency range)...")

    bands = [
        (0.1, 0.3, "Low-mid"),
        (0.2, 0.5, "Mid"),
        (0.4, 0.8, "Mid-high"),
    ]

    for low, high, name in bands:
        try:
            filtered = sg.bandpass_filter(image, low, high)

            print(f"   {name} band [{low:.1f}, {high:.1f}]:")
            print(f"     Output range: [{filtered.min():.3f}, {filtered.max():.3f}]")
            print(f"     Output std: {filtered.std():.3f}")
        except Exception as e:
            print(f"   {name} band [{low:.1f}, {high:.1f}]: Error - {e}")

    print()

    # === 8. Noise Reduction ===
    print("8. Comparing original vs smoothed for noise reduction...")

    # Apply mild low-pass to reduce noise
    denoised = sg.lowpass_filter(image, 0.4)

    print(f"   Original:")
    print(f"     Std dev: {image.std():.3f} (includes noise)")

    print(f"   Denoised (lowpass 0.4):")
    print(f"     Std dev: {denoised.std():.3f} (reduced variance)")

    # Estimate noise as difference
    noise_estimate = image - denoised
    print(f"   Estimated noise:")
    print(f"     Std dev: {noise_estimate.std():.3f}")
    print(f"     (close to injected noise std of 0.05)")

    print("\n=== Example Complete ===")
    print("\nKey observations:")
    print("- Low-pass filters smooth images by removing high frequencies (edges)")
    print("- High-pass filters emphasize edges by removing low frequencies")
    print("- Edge detection finds regions of rapid intensity change")
    print("- Sharpening enhances edges while preserving overall structure")
    print("- Band-pass filters isolate specific frequency ranges")
    print("- All operations work in frequency domain (fast for large images)")


if __name__ == "__main__":
    main()

File: examples/image_edge_detection.rs
//! Edge detection using frequency-domain filtering
//!
//! Demonstrates edge detection by applying high-pass filters in the frequency
//! domain, which emphasizes high-frequency components (edges and details).

use ndarray::Array2;
use spectrograms::SpectrogramResult;
use spectrograms::image_ops::{detect_edges_fft, highpass_filter, lowpass_filter};

fn main() -> SpectrogramResult<()> {
    println!("=== Edge Detection via Frequency Domain Filtering ===\n");

    // Create a test image with geometric shapes
    let size = 128;
    let image = create_test_image(size);

    println!("Created {}x{} test image with geometric shapes", size, size);

    // Compute image statistics
    let mean: f64 = image.iter().sum::<f64>() / (size * size) as f64;
    let variance: f64 =
        image.iter().map(|&x| (x - mean).powi(2)).sum::<f64>() / (size * size) as f64;

    println!("Original image:");
    println!("  Mean: {:.2}, Std dev: {:.2}", mean, variance.sqrt());

    // Apply high-pass filters with different cutoffs
    println!("\n=== High-Pass Filtering ===\n");

    let cutoffs = vec![0.05, 0.1, 0.2];

    for cutoff in cutoffs {
        let filtered = highpass_filter(&image.view().view(), cutoff)?;

        let filtered_mean: f64 = filtered.iter().sum::<f64>() / (size * size) as f64;
        let filtered_max = filtered.iter().map(|&x| x.abs()).fold(0.0, f64::max);

        println!("High-pass filter (cutoff: {}):", cutoff);
        println!("  Mean: {:.2e} (should be ~0)", filtered_mean);
        println!("  Max magnitude: {:.2}", filtered_max);
        println!("  Edge emphasis: {:.1}%", (filtered_max / 100.0) * 100.0);
    }

    // Compare high-pass vs low-pass
    println!("\n=== High-Pass vs Low-Pass Comparison ===\n");

    let cutoff = 0.15;
    let highpass = highpass_filter(&image.view(), cutoff)?;
    let lowpass = lowpass_filter(&image.view(), cutoff)?;

    println!("With cutoff fraction = {}:", cutoff);

    let hp_energy: f64 = highpass.iter().map(|&x| x.powi(2)).sum();
    let lp_energy: f64 = lowpass.iter().map(|&x| x.powi(2)).sum();
    let total_energy = hp_energy + lp_energy;

    println!(
        "  High-pass energy: {:.2e} ({:.1}%)",
        hp_energy,
        (hp_energy / total_energy) * 100.0
    );
    println!(
        "  Low-pass energy: {:.2e} ({:.1}%)",
        lp_energy,
        (lp_energy / total_energy) * 100.0
    );

    // Use convenience function for edge detection
    println!("\n=== Edge Detection (Convenience Function) ===\n");

    let edges = detect_edges_fft(&image.view())?;

    let edge_mean: f64 = edges.iter().sum::<f64>() / (size * size) as f64;
    let edge_max = edges.iter().map(|&x| x.abs()).fold(0.0, f64::max);

    println!("Detected edges:");
    println!("  Mean: {:.2e}", edge_mean);
    println!("  Max edge strength: {:.2}", edge_max);

    // Find strongest edges (top 1% of pixels)
    let threshold = edge_max * 0.5;
    let strong_edges = edges
        .iter()
        .filter(|&&x: &&f64| x.abs() > threshold)
        .count();

    println!(
        "  Strong edges (> {}): {} pixels ({:.1}%)",
        threshold,
        strong_edges,
        (strong_edges as f64 / (size * size) as f64) * 100.0
    );

    println!("\n=== Frequency Domain Advantages ===");
    println!("Frequency-domain edge detection:");
    println!("  * Processes entire image in parallel");
    println!("  * No directional bias (unlike Sobel, Prewitt)");
    println!("  * Adjustable frequency response");
    println!("  * Efficient for large images with FFT");

    Ok(())
}

/// Create a test image with geometric shapes (circles and rectangles)
fn create_test_image(size: usize) -> Array2<f64> {
    Array2::<f64>::from_shape_fn((size, size), |(i, j)| {
        let center = size as f64 / 2.0;
        let i_f = i as f64;
        let j_f = j as f64;

        // Circle in the center
        let dist_sq = (i_f - center).powi(2) + (j_f - center).powi(2);
        let circle_radius = size as f64 / 4.0;

        if dist_sq < circle_radius.powi(2) {
            return 100.0; // Inside circle
        }

        // Rectangle in top-left
        if i < size / 4 && j < size / 4 {
            return 80.0;
        }

        // Diagonal gradient
        let gradient = ((i + j) as f64 / (2.0 * size as f64)) * 50.0;

        // Add some noise
        let noise = ((i as f64 * 12.9898 + j as f64 * 78.233).sin() * 43758.5453).fract();

        gradient + noise * 5.0
    })
}
